### PR TITLE
Issue#316: procedure apoc.math.regr implemented

### PR DESCRIPTION
**Pull request by ArslanAle**

### CODE REVIEW

**Comment by jexp** [comment]
@ArslanAle ping ?

**Comment by ArslanAle** [comment]
@jexp yup!!

**Comment by jexp** [comment]
sorry forgot to hit "Save" on my requested changes.

**Comment by jexp** [comment]
@ArslanAle did you see my comments? Would you mind addressing them? Thanks so much.

File: src/main/java/apoc/math/Regression.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Procedure(name = "apoc.math.regr", mode = Mode.READ)
+    public Stream<Output> regr(@Name("label") String label,
+                               @Name("property_y") String y, @Name("property_x") String x ) {
+
+        SimpleRegression regr = new SimpleRegression(false);
+        Number regr_avgx = 0;
+        Number regr_avgy = 0;
+        int count = 0;
+
+        try (ResourceIterator it = db.findNodes(Label.label(label))) {
```
Comment (ArslanAle) [inline src/main/java/apoc/math/Regression.java]: @jexp what do exactly mean by generic iterator here?

File: src/main/java/apoc/math/Regression.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            while (it.hasNext()) {
+                Node node = (Node)it.next();
+                Object prop_x = node.getProperty(x, null);
```
Comment (ArslanAle) [inline src/main/java/apoc/math/Regression.java]: We have to include those points where both X and Y values are present. So we set default to null, in other to exclude the points where any of them is null.

File: src/test/java/apoc/math/RegressionTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.junit.Test;
+import org.neo4j.driver.v1.*;
+import org.neo4j.harness.junit.Neo4jRule;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author <a href="mailto:ali.arslan@rwth-aachen.de">AliArslan</a>
+ */
+public class RegressionTest {
```
Comment (ArslanAle) [inline src/test/java/apoc/math/RegressionTest.java]: @jexp would you please elaborate what is an embedded test in this context. Can you point me toward some example?

File: build.gradle [hunk 4]
```diff
@@ -84,6 +84,10 @@ dependencies {
     testCompile(group: 'org.codehaus.jackson', name: 'jackson-mapper-asl', version:'1.9.7')
     compileOnly(group: 'org.ow2.asm', name: 'asm', version:'5.0.2')
     compile group: 'com.github.javafaker', name: 'javafaker', version:'0.10'
+
+    compile group: 'org.apache.commons', name: 'commons-math3', version: '3.6.1'
+    testCompile group: 'org.neo4j.driver', name: 'neo4j-java-driver', version: '1.2.1'
```
Comment (ArslanAle) [inline build.gradle]: @jexp `org.apache.commons` is only for testing here, but `org.apache.commons` is used in procedure!!

File: src/main/java/apoc/math/Regression.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        public Number slope;
+        public Output(Number r2, Number avg_x, Number avg_y, Number slope){
+            this.r2 = r2;
+            this.avg_x = avg_x;
+            this.avg_y = avg_y;
+            this.slope = slope;
+        }
+    }
+
+    @Procedure(name = "apoc.math.regr", mode = Mode.READ)
```
Comment (ArslanAle) [inline src/main/java/apoc/math/Regression.java]: see this [https://github.com/ArslanAle/neo4j-apoc-procedures/commit/64dcb3da7314cfd97672e02a6d261e6109be5d46](commit)

File: src/main/java/apoc/math/Regression.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    public Stream<Output> regr(@Name("label") String label,
+                               @Name("property_y") String y, @Name("property_x") String x ) {
```
Comment (ArslanAle) [inline src/main/java/apoc/math/Regression.java]: see this [commit](https://github.com/ArslanAle/neo4j-apoc-procedures/commit/64dcb3da7314cfd97672e02a6d261e6109be5d46)

File: src/main/java/apoc/math/Regression.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        SimpleRegression regr = new SimpleRegression(false);
+        Number regr_avgx = 0;
```
Comment (ArslanAle) [inline src/main/java/apoc/math/Regression.java]: see this [https://github.com/ArslanAle/neo4j-apoc-procedures/commit/8afd7680ce0d064ea94903f8844a196fb6670cf2](commit)

File: src/test/java/apoc/math/RegressionTest.java (hunk 3 above)
Comment (jexp) [inline src/test/java/apoc/math/RegressionTest.java]: A test not spinning up a neo4j server but, running it against an embedded instance of Neo4j.  See: https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/3.1/src/test/java/apoc/coll/CollTest.java

**Comment by jexp** [comment]
@ArslanAle merged manually with those changes I mentioned. Thanks a lot.

Can you please all sign the CLA http://dist.neo4j.org/neo4j-cla.pdf
and send it to cla@neo4j.com and cc michael@neo4j.com ?

**Comment by ArslanAle** [comment]
@jexp thanks. I have signed and sent the CLA.

**Pull request closed without merging**